[Browser test] console.log weirdness

I added some `console.log()` calls in one storage-file-share test FileClient.ts, namely

* `    console.log("# beforeEach");` in `beforeEach()`
* `    console.log("# afterEach");` in `afterEach()`
* `console.log("# uploadRange with progress event");` in the corresponding test case.

when running browser tests. A log of `\\` got dumped onto the screen

![image.png](https://images.zenhubusercontent.com/5c51efcef73cf865af8994d5/96164f97-a3a5-4a86-a09a-f0c18ca106fb)

At the end I got V8 out-of-memory error
<details>
<--- Last few GCs --->

[22704:000002BBC5FDCF70]   111008 ms: Mark-sweep 1373.7 (1431.2) -> 1373.7 (1436.7) MB, 71.1 / 0.0 ms  (average mu = 0.110, current mu = 0.027) allocation failure scavenge might not succeed
[22704:000002BBC5FDCF70]   111085 ms: Mark-sweep 1377.0 (1436.7) -> 1376.9 (1437.2) MB, 73.0 / 0.0 ms  (average mu = 0.082, current mu = 0.053) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 000002A1F2EDC5C1]
Security context: 0x03ab0391e6e9 <JSObject>
    1: toString [000001AB9B81ED59] [buffer.js:645] [bytecode=0000012C53250811 offset=31](this=0x03eb0460ffe1 <Uint8Array map = 000001CA9CD09AA9>,encoding=0x0029c0a026f1 <undefined>,start=0x0029c0a026f1 <undefined>,end=0x0029c0a026f1 <undefined>)
    2: arguments adaptor frame: 0->3
    3: StubFrame [pc: 000002A1F2E94111]
    4: StubFrame [pc: 000002A1F2EDE...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF68FA4F04A v8::internal::GCIdleTimeHandler::GCIdleTimeHandler+5114
 2: 00007FF68FA2A0C6 node::MakeCallback+4518
 3: 00007FF68FA2AA30 node_module_register+2032
 4: 00007FF68FCB20EE v8::internal::FatalProcessOutOfMemory+846
 5: 00007FF68FCB201F v8::internal::FatalProcessOutOfMemory+639
 6: 00007FF6901D2BC4 v8::internal::Heap::MaxHeapGrowingFactor+9556
 7: 00007FF6901C9C46 v8::internal::ScavengeJob::operator=+24310
 8: 00007FF6901C829C v8::internal::ScavengeJob::operator=+17740
 9: 00007FF6901D0F87 v8::internal::Heap::MaxHeapGrowingFactor+2327
10: 00007FF6901D1006 v8::internal::Heap::MaxHeapGrowingFactor+2454
11: 00007FF68FD8CA3B v8::internal::Factory::AllocateRawWithImmortalMap+59
12: 00007FF68FD8F3AD v8::internal::Factory::NewRawOneByteString+77
13: 00007FF68FD8E519 v8::internal::Factory::NewStringFromOneByte+121
14: 00007FF68FD8E5FA v8::internal::Factory::NewStringFromUtf8+106
15: 00007FF68FCCD8FB v8::String::NewFromUtf8+363
16: 00007FF68FA184C1 node::Buffer::New+24225
17: 00007FF69022C122 v8::internal::OptimizingCompileDispatcher::Unblock+59890
18: 00007FF69022D5AD v8::internal::OptimizingCompileDispatcher::Unblock+65149
19: 00007FF69022C606 v8::internal::OptimizingCompileDispatcher::Unblock+61142
20: 00007FF69022C4EB v8::internal::OptimizingCompileDispatcher::Unblock+60859
21: 000002A1F2EDC5C1
npm ERR! code ELIFECYCLE
npm ERR! errno 134
npm ERR! @azure/storage-file-share@12.0.0 integration-test:browser: `karma start --single-run`
npm ERR! Exit status 134
npm ERR!
npm ERR! Failed at the @azure/storage-file-share@12.0.0 integration-test:browser script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\yumeng\AppData\Roaming\npm-cache\_logs\2019-12-09T20_25_40_818Z-debug.log
The script failed with exit code 134
</details>

--- a/sdk/test-utils/recorder/src/customConsoleLog.ts
+++ b/sdk/test-utils/recorder/src/customConsoleLog.ts
@@ -26,10 +26,14 @@ import { isBrowser } from "./utils";
 // - Example - console.warn("hello"); -> console.log({ warn: "hello" });
 // - Example - console.log("hello"); -> console.log({ log: "hello" });
 
-let consoleLog: (msg: any, ...args: any[]) => void;
+export let consoleLog: (msg: any, ...args: any[]) => void;
 
 if (isBrowser()) {
   consoleLog = window.console.log;
+}
+
+export function setConsoleLogForTesting(func: (msg: any, ...args: any[]) => void) {
+  consoleLog = func;
 }
 
 /**

--- a/sdk/test-utils/recorder/src/customConsoleLog.ts
+++ b/sdk/test-utils/recorder/src/customConsoleLog.ts
@@ -1,6 +1,8 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See License.txt in the project root for license information.
 
+import { isBrowser } from "./utils";
+
 // Converting content corresponding to all the console statements
 // into (JSON.stringify)-ed content in record mode for browser tests.
 //
@@ -24,12 +26,17 @@
 // - Example - console.warn("hello"); -> console.log({ warn: "hello" });
 // - Example - console.log("hello"); -> console.log({ log: "hello" });
 
+let consoleLog: (msg: any, ...args: any[]) => void;
+
+if (isBrowser()) {
+  consoleLog = window.console.log;
+}
+
 /**
  * Converts content corresponding to all the console statements into (JSON.stringify)-ed content in record mode for browser tests.
  * This allows filtering certain console.logs to generate the recordings for browser tests.
  */
 export function customConsoleLog() {
-  const consoleLog = window.console.log;
   for (const method in window.console) {
     if (
       window.console.hasOwnProperty(method) &&

--- a/sdk/test-utils/recorder/test/browser/recorder.spec.ts
+++ b/sdk/test-utils/recorder/test/browser/recorder.spec.ts
@@ -3,7 +3,7 @@ import { record, TestContextInterface, TestContext, TestContextTest } from "../.
 import xhrMock from "xhr-mock";
 import MD5 from "md5";
 import chai from "chai";
-import { consoleLog, setConsoleLogForTesting } from '../../src/customConsoleLog';
+import { consoleLog, setConsoleLogForTesting } from "../../src/customConsoleLog";
 const { expect } = chai;
 
 const expectedHttpResponse = "Hello World!";
@@ -74,9 +74,9 @@ describe("The recorder's public API, on a browser", () => {
     const savedConsoleLogParams: any[] = [];
     setConsoleLogForTesting((...params: any[]) => {
       if (params && params.length > 0) {
-        try{
-          if (JSON.parse(params[0]).writeFile){
-          savedConsoleLogParams.push(params);
+        try {
+          if (JSON.parse(params[0]).writeFile) {
+            savedConsoleLogParams.push(params);
           }
         } catch (err) {}
       }
@@ -215,14 +215,13 @@ describe("The recorder's public API, on a browser", () => {
     const savedConsoleLogParams: any[] = [];
     setConsoleLogForTesting((...params: any[]) => {
       if (params && params.length > 0) {
-        try{
-          if (JSON.parse(params[0]).writeFile){
-          savedConsoleLogParams.push(params);
+        try {
+          if (JSON.parse(params[0]).writeFile) {
+            savedConsoleLogParams.push(params);
           }
         } catch (err) {}
       }
     });
-
 
     // The recorder should start in the beforeEach call.
     // To emulate that behavior while keeping the test code as contained as possible,

--- a/sdk/test-utils/recorder/test/browser/recorder.spec.ts
+++ b/sdk/test-utils/recorder/test/browser/recorder.spec.ts
@@ -3,6 +3,7 @@ import { record, TestContextInterface, TestContext, TestContextTest } from "../.
 import xhrMock from "xhr-mock";
 import MD5 from "md5";
 import chai from "chai";
+import { consoleLog, setConsoleLogForTesting } from '../../src/customConsoleLog';
 const { expect } = chai;
 
 const expectedHttpResponse = "Hello World!";
@@ -69,11 +70,17 @@ describe("The recorder's public API, on a browser", () => {
 
     // The recorder outputs files into the console,
     // so we need to mock the console.log function to capture and test the recorder output.
-    const originalConsoleLog = console.log;
+    const originalConsoleLog = consoleLog;
     const savedConsoleLogParams: any[] = [];
-    console.log = (...params: any[]) => {
-      savedConsoleLogParams.push(params);
-    };
+    setConsoleLogForTesting((...params: any[]) => {
+      if (params && params.length > 0) {
+        try{
+          if (JSON.parse(params[0]).writeFile){
+          savedConsoleLogParams.push(params);
+          }
+        } catch (err) {}
+      }
+    });
 
     // The recorder should start in the beforeEach call.
     // We have to do this to emulate that.
@@ -97,7 +104,7 @@ describe("The recorder's public API, on a browser", () => {
     // Cleaning everything before we continue verifying the results.
     xhrMock.teardown();
     recorder.stop();
-    console.log = originalConsoleLog;
+    setConsoleLogForTesting(originalConsoleLog);
 
     // Here we confirm that the recorder generated an expected output on the console.logs.
     // This output is used to generate the recording files in the filesystem, though here we're only
@@ -203,12 +210,19 @@ describe("The recorder's public API, on a browser", () => {
     const originalXHR = XMLHttpRequest;
 
     // The recorder outputs files into the console,
-    // so we need to mock the console.log function to capture and test the recorder output.
-    const originalConsoleLog = console.log;
+    // so we need to override the consoleLog function to capture and test the recorder output.
+    const originalConsoleLog = consoleLog;
     const savedConsoleLogParams: any[] = [];
-    console.log = (...params: any) => {
-      savedConsoleLogParams.push(params);
-    };
+    setConsoleLogForTesting((...params: any[]) => {
+      if (params && params.length > 0) {
+        try{
+          if (JSON.parse(params[0]).writeFile){
+          savedConsoleLogParams.push(params);
+          }
+        } catch (err) {}
+      }
+    });
+
 
     // The recorder should start in the beforeEach call.
     // To emulate that behavior while keeping the test code as contained as possible,
@@ -234,7 +248,7 @@ describe("The recorder's public API, on a browser", () => {
     // Cleaning everything before we continue verifying the results.
     xhrMock.teardown();
     recorder.stop();
-    console.log = originalConsoleLog;
+    setConsoleLogForTesting(originalConsoleLog);
 
     // Now we check the hash has changed in the recorded console.log output.
 

